### PR TITLE
Do not rebuild project description helpers when cached

### PR DESCRIPTION
**Pull request by danibachar**
Resolves <https://github.com/tuist/tuist/issues/6992>

### Short description 📝

TL;DR see the [issue](https://github.com/tuist/tuist/issues/6992).
We faced similar issues before where we would re-build the ProjectDescriptionHelpers framework for every Tuist command.
This impose performance penalty which we like to address.

The heuristic of checking the file system to check for a hashed path artifact might caused instability. If that is indeed the case we need to address that instability, instead of compiling every time from scratch.

CC @Ernest0-Production and @fortmarek as you all have touched that code base lately

### How to test the changes locally 🧐

See steps in the issue

### Contributor checklist ✅

- [ ] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Review by fortmarek — APPROVED** [review]
Hey @danibachar 👋 
Thanks a lot for the PR. No, I don't think there was any particular reason to remove the file system check and it's something I missed. Sorry for that. Having performance smoke tests is something I'd love to cover as part of: https://community.tuist.io/t/improve-tuists-performance/130.

The PR by @Ernest0-Production was mainly meant for fixing stability during a single run and ensuring we don't recompile the same helpers simultaneously. Checking the file system shouldn't go against that.

File: Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift [hunk 1]
```diff
@@ -113,3 +158,85 @@ final class ProjectDescriptionHelpersBuilderIntegrationTests: TuistTestCase {
...
 }
+
+private class MockFileHandling: FileHandler {
+    var createFolderCount = 0
+    override public func createFolder(_: Path.AbsolutePath) throws {
+        createFolderCount += 1
+    }
+}
+
+private class MockFileSystem: FileSysteming {
```
Comment (fortmarek) [inline Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift:169]: I would avoid creating a custom mock.

If you feel strongly about mocking the file system, it should be done with [Mockable](https://github.com/Kolos65/Mockable).

File: Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift [hunk 2]
```diff
@@ -29,6 +29,51 @@ final class ProjectDescriptionHelpersBuilderIntegrationTests: TuistTestCase {
...
+                at: path,
+                projectDescriptionSearchPaths: searchPaths,
+                projectDescriptionHelperPlugins: []
+            )
+        }
+
+        // Then
+        XCTAssertEqual(Set(paths).count, 1)
+        XCTAssertEqual(fileSystem.existsCounter, 3)
+        XCTAssertEqual(fileHandler.createFolderCount, 0) // should not avoid creating a folder
```
Comment (fortmarek) [inline Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift:74]: I'd consider removing the `fileHandler` completely – that's something we will need to do anyway.

Comment (danibachar) [reply Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift:74]: Hmmm, I understand, the thing is that this is one of the verification of the test. Checking that the fileHandler.createFolder is not called.

File: Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift (hunk 1 above)
Comment (danibachar) [reply Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift:169]: Are we ok with incorporating the Mockable framework within the FileSystem SPM? I can create a PR supporting it, and then update this one to incorporate the change.
I do wonder on the choice to make the protocol conformance a struct? Should we make it a class that can be overriden?

File: Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift (hunk 2 above)
Comment (fortmarek) [reply Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift:74]: > Hmmm, I understand, the thing is that this is one of the verification of the test. Checking that the fileHandler.createFolder is not called.

Is that what we should be checking, though? Shouldn't we instead ensure the command to build the helpers is not run multiple times? Whether the folder is created or not feels like an implementation detail

File: Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift (hunk 1 above)
Comment (fortmarek) [reply Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift:169]: We can put it in the `FileSystem`, but I do wonder if we need to have the extra verification here – checking whether `createFolder` is called should be an implementation detail.

File: Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift (hunk 2 above)
Comment (danibachar) [reply Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift:74]: OK, I feel you, this should not be part of the integration tests. I can move it to the unit test level, and indeed check for the system call count like you are going in there.

I would add, that I'd still need to mock the fileSystem there such that I can return `true` for exists and early return -  going to work on that now

File: Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift (hunk 1 above)
Comment (danibachar) [reply Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift:169]: yep, you are right, see my other comment regarding the integration tests

File: Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift (hunk 2 above)
Comment (danibachar) [reply Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift]: I have changed the behaviour and moved this logic to the unit test.
Let me know if that makes sense

I will start working on adding Mockable to the file system, and replace my mock here

Comment (fortmarek) [reply Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift]: > I would add, that I'd still need to mock the fileSystem there such that I can return true for exists and early return - going to work on that now

Alternatively, you can pre-create the directory and mock the hash that determines the location instead. I think that might be better than mocking the `FileSystem`.

File: Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift [hunk 3]
```diff
@@ -173,6 +173,12 @@ public final class ProjectDescriptionHelpersBuilder: ProjectDescriptionHelpersBu
                 let dylibName = "lib\(name).dylib"
                 let modulePath = moduleCacheDirectory.appending(component: dylibName)
                 let module = ProjectDescriptionHelpersModule(name: name, path: modulePath)
+
+                // If the same helpers directory has been previously compiled, we just return the module, no need to recompile it.
+                if try await fileSystem.exists(modulePath) {
```
Comment (Ernest0-Production) [inline Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift:178]: @fortmarek Please tell me, is this correct behavior? If I'm not mistaken, the rebooting of ProjectDescriptionHelpers occurs only if the cache has been disabled. That is, if we find ourselves in this function, then we 100% always need to rebuild the helper module and not check its presence in the cache.

**Review by Ernest0-Production — CHANGES_REQUESTED** [review]

**Comment by Ernest0-Production** [comment]
@fortmarek @danibachar Hi! 
I may be wrong, but it seems that the solution is not quite suitable for solving the current problem. 

The problem of missing the cache is most likely due to the fact that `DumpService` and `GenerateService` use `ManifestLoader` instead of `CachedManifestLoader`.

File: Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift (hunk 2 above)
Comment (danibachar) [reply Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift]: As always - great feedback - fixed!

**Comment by danibachar** [comment]
> @fortmarek @danibachar Hi! I may be wrong, but it seems that the solution is not quite suitable for solving the current problem.
> 
> The problem of missing the cache is most likely due to the fact that `DumpService` and `GenerateService` use `ManifestLoader` instead of `CachedManifestLoader`.

Oh that is interesting, I was not aware of the two loaders and their difference.
I can definitely change the service to use the cache manifest loader, but why would any service not use it to begin with?

**Comment by danibachar** [comment]
> #6992

See - https://github.com/tuist/tuist/pull/7003

File: Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift (hunk 3 above)
Comment (fortmarek) [reply Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift]: I don't think that's true. I do think we should use `CachedManifestLoader` in the `Dump` and `Generate` services but it's still possible that a manifest's hash has changed _but_ the project description helpers have not. In those cases, we can still skip building the project description helpers.

**Review by fortmarek — APPROVED** [review]
Let me merge this one as well as we should skip building project description helpers if they are already built on the disk
